File: backend/app/translation/adapters/mock.py
from __future__ import annotations

import asyncio
import time

from app.translation.ports import (
    LLMStreamEvent,
    ModelCapabilities,
    StreamDelta,
    StreamFinal,
    StreamStart,
    TranslationLLM,
    TranslationRequest,
)


class MockTranslationAdapter(TranslationLLM):
    provider = "mock"
    model = "mock-sonnet"
    capabilities = ModelCapabilities(
        streaming=True,
        glossary=True,
        json_mode=False,
        prompt_caching=False,
        max_context_tokens=8_000,
    )

    async def translate_stream(self, req: TranslationRequest):
        started = time.perf_counter()
        yield StreamStart(provider=self.provider, model=self.model)
        await asyncio.sleep(0)
        translated = f"[{req.target_lang}] {req.text}"
        halfway = max(1, len(translated) // 2)
        yield StreamDelta(text=translated[:halfway])
        await asyncio.sleep(0)
        yield StreamDelta(text=translated[halfway:])
        total_ms = int((time.perf_counter() - started) * 1000)
        yield StreamFinal(
            text=translated,
            finish_reason="stop",
            latency_first_token_ms=1,
            latency_total_ms=total_ms,
        )

File: go-app/internal/translation/adapters/mock.go
package adapters

import (
	"fmt"
	"time"

	"talk/go-app/internal/translation"
)

type MockTranslationAdapter struct{}

func (a *MockTranslationAdapter) Provider() string { return "mock" }
func (a *MockTranslationAdapter) Model() string    { return "mock-sonnet" }
func (a *MockTranslationAdapter) Capabilities() translation.ModelCapabilities {
	return translation.ModelCapabilities{Streaming: true, Glossary: true}
}
func (a *MockTranslationAdapter) TranslateStream(req translation.TranslationRequest) ([]translation.StreamEvent, error) {
	translated := fmt.Sprintf("[%s] %s", req.TargetLang, req.Text)
	halfway := len(translated) / 2
	if halfway < 1 {
		halfway = 1
	}
	total := int(time.Millisecond)
	first := 1
	return []translation.StreamEvent{
		translation.StreamStart{Provider: a.Provider(), Model: a.Model()},
		translation.StreamDelta{Text: translated[:halfway]},
		translation.StreamDelta{Text: translated[halfway:]},
		translation.StreamFinal{Text: translated, FinishReason: "stop", LatencyFirstTokenMS: &first, LatencyTotalMS: &total},
	}, nil
}
